Move Codegen pass pipelines to nest on `FunctionOpInterface`.

MaheshRavishankar opened the pull request:
This PR modifies the codegen backends to have the lowering pass
pipelines nest on `FunctionOpInterface`. This allows running different
pass pipelines on functions within the dispatch. This would allow you
to have things like

```
func.func @foo_pipeline(...) {

}

func.func @bar_pipeline(...) {

}

func.func @entry_point() {
  if (<condnn for foo pipeline based lowering>) {
    foo_pipeline()
  } else {
    bar_pipeline()
  }
}
```

To connect everything the following things are done

1) The `iree_codegen.translation_info` attribute that was set on entry
   point operations are now set on the surrounding function. This
   allows implementing a lowering strategy on a function.

2) The GPU backends set the `workgroup_size` and `subgroup_size` on
   the `hal.executable.export` operation. To unwind this, the
   `translation_info` has fields for `workgroup_size` and
   `subgroup_size`. This allows GPU backends to set the expected
   `workgroup_size` and `subgroup_size` on the `translation_info`
   itself (which is now on the surrounding function).

3) A pass is added after lower strategies to
   `ReconcileTranslationInfo`. The intent of this pass is to take the
   `translation_info` on each function and set the values for
   `workgroup_size` and `subgroup_size` on the
   `hal.executable.export`. Eventually this would also be a place
   where the number of workgroups is populated on the
   `hal.executable.export` (instead of doing it on
   `TileAndDistributeToWorkgroups` as it is done today).

4) All backends `*SelectLoweringStrategy` work as Module pass. These
   need to be Module passes since transform dialect tends to inject
   the transform script within the module.

5) The `*LowerExecutableStrategy` works at `FunctionOpInterface` now.

6) The transform dialect interpreter has to run on `Module`
   granularity, so a new pass `LowerExecutableUsingTransformDialect`
   is added. This runs the transform interpreter before
   `*SelectLoweringStrategy`. After this pass is run, the
   `translation_info` is expected to have the pipeline be set to
   `None` to skip subsequent lowering pipelines.

7) Most tests are now moved to remove the boiler plates surrounding
   `hal.executable` and `hal.executable.variant`.

This does most of the heavy lifting for running lowering strategies
per function-like op. The biggest missing piece are

1) The `TileAndDistributeOnWorkgroups` ops still cannot really be run
   on a dispatch with multiple functions since it updates the
   `hal.executable.export`. To address this, the pass will have to
   move to use `scf.forall`.

2) Some optimizations expect static workgroup count. Those currently
   go upto the `hal.executable.export` op to get these values (that
   were populated by `TileAndDistributeToWorkgroups`). When moving to
   `scf.forall` this will be available withint the function.


ci-extra: build_test_all_arm64, build_test_all_windows, build_test_all_maxos_arm64, build_test_all_macos_x86_64, test_nvidia_a100

hanhanW commented on compiler/src/iree/compiler/Codegen/Common/TileSizeSelection.cpp:
```diff
@@ -101,6 +101,9 @@ SmallVector<int64_t> TilingConfig::getFusableLevels() {
     return {0};
+  case 3:
+    // Only distribution level.
+    return {0, 1};
```
This is correct, but where is it from? There are too many diffs in the link you shared (i.e, https://github.com/openxla/iree/pull/16665/files#diff-7d0ef2ec4b800099c356491a7e7f12486452befb8a7b02a1354c1cf088ae2c4bR367).

I think we probably want to update the documentations. IIUC, it is the `[[distribution], [vector-common-parallel], [vector-reduction]]` case, which is not listed in the doc..

https://github.com/openxla/iree/blob/e913ff1d28c44afa6eb9ed88e938f5ace32655b1/compiler/src/iree/compiler/Codegen/Common/TileSizeSelection.h#L22-L29

https://github.com/openxla/iree/blob/e913ff1d28c44afa6eb9ed88e938f5ace32655b1/compiler/src/iree/compiler/Codegen/Common/TileSizeSelection.cpp#L21-L30

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:
```diff
@@ -533,282 +494,271 @@ void addMmt4dTilingExpertPassPipeline(OpPassManager &passManager,
+      .addPass(createMemrefCopyToLinalgPass)
+      .addPredicatedPass(clCheckLinalgVectorization,
+                         createLLVMCPUEmitVectorizationRemarksPass)
+
```
nit: remove the empty line

hanhanW replied:
bump the comment

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:
```diff
@@ -533,282 +494,271 @@ void addMmt4dTilingExpertPassPipeline(OpPassManager &passManager,
+      .addPass(createHoistStaticallyBoundAllocationsPass)
+
+      // Use `arith.minf/maxf` instead of `arith.minimumf/maximumf`.
+      .addPredicatedPass(clUseFastMinMaxOps, createReplaceSlowMinMaxOpsPass);
```
nit: remove blank lines between `addPass`. That's also the style we're following in Flow folders. E.g.,

https://github.com/openxla/iree/blob/e913ff1d28c44afa6eb9ed88e938f5ace32655b1/compiler/src/iree/compiler/Dialect/Flow/Transforms/Passes.cpp#L141-L166

hanhanW reviewed (commented):
The changes on CPU side look okay to me. So the translation_info will be set on func.func ops?

MaheshRavishankar commented on compiler/src/iree/compiler/Codegen/Common/MaterializeUserConfigs.cpp:
```diff
@@ -112,115 +108,86 @@ struct MaterializeUserConfigsPass
-    // From now on, we know we have a transform dialect strategy. We now need to
-    // ensure it can resolve and apply in a subsequent interpreter pass or else
-    // we need to fall back to codegen.
-    bool failedToResolve = false;
```
@qedawkins please take a look at this and confirm that removing this is OK.

qedawkins replied:
Can take a closer look on Wednesday, but this looks fine to remove. I'm wondering why we're changing the nesting of this pass though. I would expect it to keep either a module or variant scope.

MaheshRavishankar replied:
I am not sure either. Looking at a code, it works as a function pass... so I changed it. I also punted on how this all interacts with transform dialect (but I am looking at that issue now). It makes sense to keep this as a module or variant pass (ideally the former). Ill keep a pointer to change this as a module pass... but as written it doesnt need to be.

qedawkins replied:
Sounds good. We don't care yet so nesting on the function for now is fine, but I'm assuming this pass (and the pass that runs kernel config) will need to nest on the module eventually. Wrt your first question then, if this is nesting on functions then it should be fine to remove. I can help do a full review of this pr once I'm back too.

qedawkins replied:
ok rereading this code, I think the scope exit might be required to make the `TransformStrategies` work.

MaheshRavishankar committed:
Move Codegen pass pipelines to nest on `FunctionOpInterface`.

This PR modifies the codegen backends to have the lowering pass
pipelines nest on FunctionOpInterface. This allows running different
pass pipelines on functions within the dispatch. This would allow you
to have things like

```
func.func @foo_pipeline(...) {

}

func.func @bar_pipeline(...) {

}

func.func @entry_point() {
  if (<condnn for foo pipeline based lowering>) {
    foo_pipeline()
  } else {
    bar_pipeline()
  }
}
```
To connect everything the following things are done

1) The `iree_codegen.translation_info` attribute that was set on entry
   point operations are now set on the surrounding function. This
   allows implementing a lowering strategy on a function.

2) The GPU backends set the `workgroup_size` and `subgroup_size` on
   the `hal.executable.export` operation. To unwind this, the
   translation_info has fields for `workgroup_size` and
   `subgroup_size`. This allows GPU backends to set the expected
   `workgroup_size` and `subgroup_size` on the `translation_info`
   itself (which is now on the surrounding function).

3) A pass is added after lower strategies to
   `ReconcileTranslationInfo`. The intent of this pass is to take the
   `translation_info` on each function and set the values for
   `workgroup_size` and `subgroup_size` on the
   `hal.executable.export`. Eventually this would also be a place
   where the number of workgroups is populated on the
   `hal.executable.export` (instead of doing it on
   `TileAndDistributeToWorkgroups` as it is done today).

4) All backends `*SelectLoweringStrategy` work as `Module` pass. These
   need to be Module passes since transform dialect tends to inject
   the transform script within the module.

5) The `*LowerExecutableStrategy` works at `FunctionOpInterface` now.

6) The transform dialect interpreter has to run on `Module` granularity,
   so a new pass `LowerExecutableUsingTransformDialect` is added. This
   runs the transform interpreter before
   `*SelectLoweringStrategy`. After this pass is run, the
   translation_info is expected to have the pipeline be set to None to
   skip subsequent lowering pipelines.

7) Most tests are now moved to remove the boiler plates surrounding
   `hal.executable` and `hal.executable.variant`.

This does most of the heavy lifting for running lowering strategies
per function-like op. The biggest missing piece are

1) The `TileAndDistributeOnWorkgroups` ops still cannot really be run
   on a dispatch with multiple functions since it updates the
   `hal.executable.export`. To address this, the pass will have to move
   to use scf.forall.

2) Some optimizations expect static workgroup count. Those currently
   go upto the `hal.executable.export` op to get these values (that were
   populated by `TileAndDistributeToWorkgroups`). When moving to
   `scf.forall` this will be available withint the function.

benvanik commented on compiler/src/iree/compiler/Dialect/HAL/IR/HALOps.td:
```diff
@@ -2095,6 +2095,15 @@ def HAL_ExecutableExportOp : HAL_Op<"executable.export", [
         Location loc, Value device, ValueRange workload, OpBuilder &builder);
+
+    // Helper function to return the subgroup size.
+    std::optional<size_t> getSubgroupSizeAsSizeT() {
```
this is an invalid type cast - getZExtValue returns uint64_t, not size_t, and size_t != uint64_t.
prefer `getSubgroupSizeAsUInt` instead and return a uint64_t to match what MLIR commonly does (ala IntegerAttr::getUInt) - or `getSubgroupSizeAsInt` and int64_t (as is more common in MLIR)

MaheshRavishankar replied:
yeah sounds good. Will change it.

benvanik reviewed (approved):
![image](https://github.com/openxla/iree/assets/75337/35160dea-89ff-4707-9cec-c0c659b85779)
![giphy 1](https://github.com/openxla/iree/assets/75337/791dc748-2483-4985-8f97-46c1d814fdd2)

lots of really nice cleanup here and it sets the foundation for a lot of fun future work!
LGTM besides one nit but I mostly skimmed the Codegen/ files and would want hanhan's approval there

MaheshRavishankar committed:
Fix wrong escape condition in RemoveSingleIterationLoop.

hanhanW commented on compiler/src/iree/compiler/Codegen/Common/LowerExecutableUsingTransformDialect.cpp:
Do we have a test for this new pass?

MaheshRavishankar replied:
Not specifically for this, but it is tested in some LLVMGPU passes (like `transform_dialect_distribute_forall.mlir`). It is just a pass to invoke other passes. So I think this is enough.

hanhanW commented on compiler/src/iree/compiler/Codegen/Common/ReconcileTranslationInfo.cpp:
ditto: do we have a test for this new pass?

MaheshRavishankar replied:
Added (will show up when I push).

hanhanW commented on compiler/src/iree/compiler/Codegen/Common/ReconcileTranslationInfo.cpp:
```diff
@@ -0,0 +1,158 @@
+static FailureOr<int64_t> reconcileSubgroupSize(
+    ArrayRef<IREE::Codegen::TranslationInfoAttr> translationInfos) {
+  if (translationInfos.empty()) {
+    return int64_t();
```
Do we just return `0` here?

(This is interesting to me because this is my first time seeing it. I'm 99% sure that it is 0. I did run a test locally to verify it. :P )

MaheshRavishankar replied:
I was just trying to see what the tablegen generated code does for `DefaultValuedAttr` and did the same to check for / return default values.

kuhar replied:
> (This is interesting to me because this is my first time seeing it. I'm 99% sure that it is 0. I did run a test locally to verify it. :P )

it is, the umbrella term is value initialization https://en.cppreference.com/w/cpp/language/value_initialization, which ends up being zero-initialization for scalars.

hanhanW commented on compiler/src/iree/compiler/Codegen/Common/ReconcileTranslationInfo.cpp:
```diff
@@ -0,0 +1,158 @@
+    // lowering configs and translation infos will be deleted.
+    std::optional<int64_t> wavesPerEu = getWavesPerEu(translationInfo);
+    if (wavesPerEu) {
+      funcOp->setAttr("waves_per_eu", IntegerAttr::get(IndexType::get(context),
+                                                       wavesPerEu.value()));
```
I'm seeing many `XXX::get` in the pass. I think we can create a `Builder builder(context)` and get the attributes from it. In this case, we can do `builder.getIndexAttr(wavesPerEu.value())` here.

MaheshRavishankar replied:
Dont want to create a builder for attributes. I want to make sure there is no IR modification in this pass (at least for now, will change in future). So Ill leave this as is.

hanhanW replied:
It is a Builder but not OpBuilder. So I think it fits what you want?

hanhanW replied:
This builder only has methods for getting attributes and types.

https://github.com/llvm/llvm-project/blob/a9d4ddd98a0bc495126027122fdca751b6841ceb/mlir/include/mlir/IR/Builders.h#L50-L205

hanhanW commented on compiler/src/iree/compiler/Codegen/Common/ReconcileTranslationInfo.cpp:
```diff
@@ -0,0 +1,158 @@
+        return IntegerAttr::get(IndexType::get(context), value);
+      });
+  exportOp.setWorkgroupSizeAttr(
+      ArrayAttr::get(context, workgroupSizeArrayAttr));
```
If we have the `Builder` we can simplify it with SmallVector like:

```cpp
SmallVector<int64_t> workgroupSize(3, 1);
for (...) {
 ...
}
exportOp.setWorkgroupSizeAttr(builder.getI64ArrayAttr(workgroupSize));
```

https://github.com/llvm/llvm-project/blob/a9d4ddd98a0bc495126027122fdca751b6841ceb/mlir/include/mlir/IR/Builders.h#L168

MaheshRavishankar replied:
Good point, but see above. I want to avoid builders for now.

hanhanW commented on compiler/src/iree/compiler/Codegen/Common/ReconcileTranslationInfo.cpp:
```diff
@@ -0,0 +1,158 @@
+  }
+  if (reconciledSubgroupSize.value() != int64_t()) {
+    exportOp.setSubgroupSizeAttr(IntegerAttr::get(
+        IndexType::get(context), reconciledSubgroupSize.value()));
```
This can be simpler if we use `Builder`, e.g., `builder. getIndexAttr(reconciledSubgroupSize.value())`.

MaheshRavishankar replied:
ditto

hanhanW commented on compiler/src/iree/compiler/Codegen/Common/TileAndDistributeToWorkgroupsPass.cpp:
```diff
@@ -281,24 +281,18 @@ struct TileAndDistributeToWorkgroupsPass
+  auto funcOp = getOperation();
 
-    Block *body = exportOp.getWorkgroupCountBody();
+  // TODO(MaheshRavishankar): THe logic of lowering workgroup count
```
nit: The

```suggestion
  // TODO(MaheshRavishankar): The logic of lowering workgroup count
```

hanhanW commented on compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenAttrs.cpp:
```diff
@@ -59,10 +59,22 @@ ArrayAttr ExportConfigAttr::getWorkgroupSizeIndexArray() {
       DispatchLoweringPassPipelineAttr::get(context, passPipeline);
-  return get(context, pipelineAttr, codegenSpec, configuration);
+  return get(context, pipelineAttr, codegenSpec, workgroupSize,
+             subgroupSize.value_or(int64_t()), configuration);
```
ditto: maybe just use `value_or(0)` instead of `int64_t`.

MaheshRavishankar replied:
(see above. Just want to be consistent with what the tablegen generated code does)

hanhanW commented on compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenAttrs.cpp:
```diff
@@ -59,10 +59,22 @@ ArrayAttr ExportConfigAttr::getWorkgroupSizeIndexArray() {
+  auto pipelineAttr =
+      DispatchLoweringPassPipelineAttr::get(context, passPipeline);
+  return get(context, pipelineAttr, /*codegenSpec=*/SymbolRefAttr(),
+             workgroupSize, subgroupSize.value_or(int64_t()), configuration);
```
ditto: maybe just `0`?

MaheshRavishankar replied:
ditto :)

hanhanW commented on compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenAttrs.cpp:
```diff
@@ -324,80 +347,46 @@ namespace mlir::iree_compiler {
-  if (!workgroupSize.empty()) {
-    exportOp->setWorkgroupSizeAttr(getIndexArrayAttr(context, workgroupSize));
+std::optional<int64_t> getSubgroupSize(FunctionOpInterface funcOp) {
+  IREE::Codegen::TranslationInfoAttr translationInfo =
+      getTranslationInfo(funcOp);
```
nit: it is already within `IREE::Codegen` namespace scope, so we don't need the prefix. It looks cleaner and it saves us 1 line.

hanhanW reviewed (commented):
Really nice cleanup!

(And thanks for the review tips. I also hide whitespaces during the review. Please hold the PR a bit until we land the LLVM commit to upstream. :))

![image](https://github.com/openxla/iree/assets/6949191/ed7e212b-f6de-470f-bb36-a1df372b8ab0)

MaheshRavishankar committed:
Use transform dialect lowering pass on CPU backend.

MaheshRavishankar committed:
Create new pass for SPIR-V transform dialect usage.

MaheshRavishankar committed:
Add debug prints for LLVM CPU pass pipeline and fix matmul test generator.

qedawkins commented on compiler/src/iree/compiler/Codegen/Common/GPU/GPUCheckResourceUsage.cpp:
```diff
@@ -104,21 +104,20 @@ static LogicalResult checkGPUAllocationSize(
+  auto funcOp = getOperation();
+  unsigned limit = this->getSharedMemoryLimit
+                       ? this->getSharedMemoryLimit(funcOp)
+                       : 64 * 1024;
```
self note: We need to start documenting where magic constants like this are coming from.

qedawkins commented on compiler/src/iree/compiler/Codegen/Common/GPU/Passes.h:
```diff
@@ -136,6 +137,8 @@ createGPUVectorAlloc();
 // current function containing those vector ops as the argument.
 // If nullptr, warp size 32 will be used.
+// TODO: This kind of call back function is a really really bad idea
+// This should be easier to resolve than doing this.
```
note: IIRC this is an artifact of codegen differences between SPIR-V and LLVMGPU. We should try to clean this up once we get a chance to do the ROCDL/NVVM split.

qedawkins commented on compiler/src/iree/compiler/Codegen/Common/LowerExecutableUsingTransformDialect.cpp:
```diff
@@ -0,0 +1,71 @@
+  if (funcOps.empty() || !llvm::hasSingleElement(funcOps)) {
+    // Can only handle dispatches with single functions on the transform dialect
+    // path.
+    return;
```
nit: `signalPassFailure`

MaheshRavishankar replied:
I think Ill leave it as that for now... for function list being empty, this pass should be a no-op. Even when we add support for multiple functions, this pass is run before all the lower executables pass. It should just do nothing. So silent return is better.

qedawkins replied:
yeah that's fine. I thought this was run only for transform dialect cases until you explained it to me.

qedawkins commented on compiler/src/iree/compiler/Codegen/Common/LowerExecutableUsingTransformDialect.cpp:
```diff
@@ -0,0 +1,71 @@
+  if (!translationInfo || translationInfo.getDispatchLoweringPassPipeline() !=
+                              IREE::Codegen::DispatchLoweringPassPipeline::
+                                  TransformDialectCodegen) {
+    return;
```
nit: `signalPassFailure`

MaheshRavishankar replied:
Same here. This isnt a failure. This is saying that the lowering isnt using transform dialect. Just return.

qedawkins commented on compiler/src/iree/compiler/Codegen/Common/RemoveTrivialLoops.cpp:
```diff
@@ -111,14 +113,16 @@ class RemoveSingleIterationLoopPass final
-    SmallVector<int64_t> workgroupSize = getWorkgroupSize(*exportOp);
+    std::optional<SmallVector<int64_t>> workgroupSize =
+        getWorkgroupSize(funcOp);
+    if (!workgroupSize) {
```
This should be rewritten to use value bounds interface + a custom interface function for thread/block id/dim.

MaheshRavishankar replied:
This is long standing legacy. Once I move things to use scf.forall, these might not be needed.

qedawkins commented on compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenAttrs.cpp:
```diff
@@ -89,6 +102,15 @@ LogicalResult TranslationInfoAttr::verify(
+  }
+  if (subgroupSize < 0) {
+    return emitError() << "subgroup size value cannot be negative";
+  }
```
One thought, we could make workgroup size + subgroup size its own attribute class rather than hard coding them into the translation info, then either attach it to the configuration dict, or carve out a generic field for required things (where the configuration includes optional things).

MaheshRavishankar replied:
I think given that these are values on the hal.executable.export, I added them here to force all backends to think about it even if they dont set the workgroup size or subrgroup size.

ScottTodd commented on compiler/src/iree/compiler/Codegen/LLVMGPU/test/reduction_pipeline.mlir:
There are a few new empty test files in this PR. Did you intend on filling those in with new tests, or are they leftover from outdated code?

* `compiler/src/iree/compiler/Codegen/LLVMGPU/test/conv_pipeline.mlir`
* `compiler/src/iree/compiler/Codegen/LLVMGPU/test/reduction_pipeline.mlir`
* `compiler/src/iree/compiler/Codegen/LLVMGPU/test/transform_dialect_forall.mlir`

MaheshRavishankar replied:
Oh good catch. Thanks! Removed these. Artifacts of scripts I was using to do some test changes.

MaheshRavishankar committed:
Add lit tests for translation info illegal spec and reconcile translation info.

MaheshRavishankar committed:
Address review comments.

qedawkins commented on compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUVectorDistribute.cpp:
```diff
@@ -362,7 +362,17 @@ struct LLVMGPUVectorDistributePass
         workgroupSize[i] = llvm::cast<IntegerAttr>(size).getInt();
       }
     } else {
```
self note: The logic to get the workgroup size from the function here was a placeholder prior to this PR to avoid needing to specify the whole variant in tests. Drop this in a follow up.

MaheshRavishankar replied:
Are there tests that do this... I can change them right now.

qedawkins replied:
All of these ones: https://github.com/openxla/iree/blob/main/compiler/src/iree/compiler/Codegen/LLVMGPU/test/vector_distribute_layout.mlir + https://github.com/openxla/iree/blob/main/compiler/src/iree/compiler/Codegen/LLVMGPU/test/vector_distribute_conversion.mlir

I don't care either way if it's addressed here or in a follow up though.

qedawkins reviewed (approved):
I don't see anything to block on. I left a few notes for things to follow up on.

kuhar commented on compiler/src/iree/compiler/Codegen/Common/PassUtils.cpp:
```diff
@@ -0,0 +1,23 @@
+      .Case<func::FuncOp, IREE::Util::FuncOp>([&](auto funcOp) {
+        return OpPassManager(funcOp.getOperationName());
+      });
+  return std::nullopt;
```
This is unreachable. Did you intend to add a default case instead?

MaheshRavishankar replied:
Good catch!

kuhar commented on compiler/src/iree/compiler/Codegen/Common/ReconcileTranslationInfo.cpp:
```diff
@@ -0,0 +1,158 @@
+    return int64_t();
+  }
+  int64_t subgroupSize = translationInfos.front().getSubgroupSize();
+  for (auto translationInfo : translationInfos.drop_front()) {
```
nit: I prefer to have explicit types in places like this, where it's not obvious based on the RHS only. Same everywhere below

MaheshRavishankar replied:
I think you can use `auto` when type is obvious from the context. `translationInfos` type is just above (`ArrayRef<TranslationInfoAttr>`) so this is `TranslationInfoAttr`.

kuhar replied:
I think this boils down what your definition of obvious is. For me is just by looking at the RHS.

kuhar commented on compiler/src/iree/compiler/Codegen/Common/ReconcileTranslationInfo.cpp:
```diff
@@ -0,0 +1,158 @@
+    return signalPassFailure();
+  }
+  std::array<int64_t, 3> workgroupSize = {1, 1, 1};
+  for (auto [index, size] : llvm::enumerate(reconciledWorkgroupSize.value())) {
+    workgroupSize[index] = size;
+  }
```
nit: this could be an `llvm::copy`

MaheshRavishankar replied:
Yeah maybe, but I dont understand the API completely.

kuhar commented on compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenAttrs.cpp:
```diff
@@ -324,80 +347,46 @@ namespace mlir::iree_compiler {
-    exportOp->setSubgroupSizeAttr(Builder(context).getIndexAttr(*subgroupSize));
+  // The underlying storage sets 0 to optional scalar integer value. So if set
+  // to 0, return as not set.
+  if (translationInfo.getSubgroupSize() == int64_t()) {
```
```suggestion
  if (translationInfo.getSubgroupSize() == 0) {
```
?

MaheshRavishankar replied:
I keeping everywhere I check default to be consistent with what tablegen generates. They are the same. So Ill leave it as is.

kuhar commented on compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/test/lowering_config_attr.mlir:
```diff
@@ -49,15 +49,13 @@ module {
 // CHECK: #config = #iree_codegen.lowering_config<tile_sizes = []>
-// CHECK: #translation = #iree_codegen.translation_info<CPUDefault>
-// CHECK: #compilation = #iree_codegen.compilation_info<lowering_config = #config, translation_info = #translation, workgroup_size = [16, 4, 1], subgroup_size = 32>
+// CHECK: #translation = #iree_codegen.translation_info<CPUDefault workgroup_size = [16, 4, 1] subgroup_size = 32>
```
I'm surprised we don't have any commas between these

MaheshRavishankar replied:
adding the `,` makes the parser ambiguous. The parser is not a look ahead parser. So if I do the assembly format
```
(`,` `workgroup_size` `=` ...)? (`,` `subgroup_size` `=`)
```

on parsing the `,` it doesnt know whether it is parsing workgroup size or subgroup size. So I just dropped it

kuhar commented on compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPULowerExecutableTarget.cpp:
```diff
@@ -61,75 +62,38 @@ class LLVMCPULowerExecutableTargetPass
-        "unhandled compilation of entry point functions with different "
-        "translation info");
-    return signalPassFailure();
+  auto funcOp = getOperation();
+  auto target = IREE::HAL::ExecutableTargetAttr::lookup(funcOp);
```
Also here, I'd prefer to have the types instead of  `auto`. I think this hurts readability outside of IDE,

MaheshRavishankar replied:
The `IREE::HAL::ExecutableAttr::lookup` returns a element of that type. It is lile `rewriter.create<Op>` creates a `Op`.

MaheshRavishankar committed:
Make `MaterializeUserConfig` pass a module level pass.

MaheshRavishankar committed:
Rebase and address reviewer comments.

MaheshRavishankar committed:
Fix linter errors.

MaheshRavishankar committed:
Fix newly added test.

MaheshRavishankar committed:
Fix build error.

MaheshRavishankar committed:
More build fixes
